src/core/IronPython.Modules/signal.PosixSignalState.cs

```diff
@@ -0,0 +1,102 @@
+using IronPython.Runtime.Operations;
+using System.Threading.Tasks;
+using Microsoft.Scripting.Runtime;
+using Microsoft.Scripting.Hosting.Shell;
+
+namespace IronPython.Modules {
+    public static partial class PythonSignal {
+
+        [SupportedOSPlatform("linux")]
+        [SupportedOSPlatform("macos")]
+        private class PosixSignalState : PythonSignalState {
+            private readonly PosixSignalRegistration?[] _signalRegistrations;
+            private ConsoleCancelEventHandler? _consoleHandler;
+
+
+            public PosixSignalState(PythonContext pc) : base(pc) {
+                _signalRegistrations = new PosixSignalRegistration[NSIG];
+            }
+
+
+            protected override void Dispose(bool disposing) {
+                if (disposing) {
+                    Array.ForEach(_signalRegistrations, reg => reg?.Dispose());
+                    Array.Clear(_signalRegistrations, 0, _signalRegistrations.Length);
+                }
+                base.Dispose(disposing);
+            }
+
+
+            public override void SetPyHandler(int signalnum, object value) {
+                // normalize handler reference
+                if (value is int i) {
+                    value = (i == SIG_IGN) ? sig_ign : sig_dfl;
+                }
+
+                if (TryGetPyHandler(signalnum, out object? existingValue) && ReferenceEquals(existingValue, value)) {
+                    // no change
+                    return;
+                }
+
+                base.SetPyHandler(signalnum, value);
+
+                if (signalnum == SIGINT) {
+                    // SIGINT is special, we need to disable the handler on the console so that we can handle Ctrl+C here
+                    if (DefaultContext.DefaultPythonContext.Console is BasicConsole console) {
+                        if (!ReferenceEquals(value, default_int_handler)) {
+                            // save the console handler so it can be restored later if necessary
+                            _consoleHandler ??= console.ConsoleCancelEventHandler;
+                            // disable the console handler
+                            console.ConsoleCancelEventHandler = null;
+                        } else if (_consoleHandler != null) {
+                            // default_int_handler: restore the console handler, which is de facto default_int_handler
+                            console.ConsoleCancelEventHandler = _consoleHandler;
+                            _consoleHandler = null;
+                        }
+                    }
+                }
+
+                // remember to unregister any previous handler
+                var oldReg = _signalRegistrations[signalnum];
+
+                if (ReferenceEquals(value, sig_dfl)) {
+                    _signalRegistrations[signalnum] = null;
+                } else if (ReferenceEquals(value, sig_ign)) {
+                    _signalRegistrations[signalnum] = PosixSignalRegistration.Create((PosixSignal)signalnum,
```

Comment:
Does the `PosixSignal` enum match up with the `signalnum` values? Looking at the docs the .NET enum values are negative while the `signalnum` are positive?

Reply:
Yes, the enum values are negative and platform-independent, the positive values are actual platform-dependent signal numbers. Platform-dependent signal numbers is what CPython uses (at least on POSIX, I am still a little bit puzzled by what is going on on Windows).

I think it is a better scheme than what Mono.Unix uses with its enum values, since it is future-proof (the enum values do not cover, nor do they need to cover all available signals on a given platform).